src/lua/protobuf_wireformat.lua

```diff
@@ -97,6 +141,29 @@ local function encode_varint(value)
     return ffi.string(buf, size)
 end
 
+-- Decode input value as VARINT without a tag.
+--
+-- Input value type: string
+--
+-- This is a helper function to decode tag and data values.
+local function decode_varint(ctx)
+    local msb
+    local res = 0
+    local len = 0
+    repeat
+        local payload = ffi.cast('uint64_t', string.byte(ctx.val, ctx.pos))
```

Comment:
It seems, here we can read data out of the buffer boundary. Let's add a boundary check. Consider introducing functions to read data from the current position of the decoding context: read one byte as a number, read 4 bytes as cdata<uint32_t> and so on -- anything that you really need. These functions should perform the boundary checks. (And also this way it is easier to test different buffer implemetations in a future to choose the most performant.)

Reply:
As mentioned above added a function `read_bytes` that read needed amout of bytes from decoding context with boundary check.

Reply:
I think we can reduce amount of allocated GC objects (substrings), but I'm OK to go with this simple solution for now. At least, we're doing a very similar thing in encoding.